// testcafe/lib/server/server_errs.js
exports.TESTCAFE_CONFIG_NOT_SPECIFIED = "TESTCAFE_CONFIG_NOT_SPECIFIED";

exports.API_TESTS_TARGET_NOT_EXIST = "API_TESTS_TARGET_NOT_EXIST";

exports.API_UNKNOWN_FIXTURE_TYPE = "API_UNKNOWN_FIXTURE_TYPE";

exports.API_NO_TESTS_TO_RUN = "API_NO_TESTS_TO_RUN";

exports.API_TASK_WORKER_DOESNT_EXIST = "API_TASK_WORKER_DOESNT_EXIST";

exports.API_NO_WORKERS_SPECIFIED_FOR_TASK = "API_NO_WORKERS_SPECIFIED_FOR_TASK";

exports.REPORTER_WRITE_REPORT_FAILED = "REPORTER_WRITE_REPORT_FAILED";

exports.CONFIG_READ_FILE_FAILED = "CONFIG_READ_FILE_FAILED";

exports.CONFIG_PARSING_FAILED = "CONFIG_PARSING_FAILED";

exports.CONFIG_PROPERTY_IS_NOT_STRING = "CONFIG_PROPERTY_IS_NOT_STRING";

exports.CONFIG_PROPERTY_IS_REQUIRED = "CONFIG_PROPERTY_IS_REQUIRED";

exports.CONFIG_PROPERTY_CAN_NOT_BE_EMPTY = "CONFIG_PROPERTY_CAN_NOT_BE_EMPTY";

exports.CONFIG_UNSUPPORTED_PROPERTY = "CONFIG_UNSUPPORTED_PROPERTY";

exports.CONFIG_PORT_VALUE_IS_NOT_NUMBER = "CONFIG_PORT_VALUE_IS_NOT_NUMBER";

exports.CONFIG_PORT_VALUE_IS_OUT_OF_RANGE = "CONFIG_PORT_VALUE_IS_OUT_OF_RANGE";

exports.CONFIG_INVALID_PROPERTY_FORMAT = "CONFIG_ERR_WRONG_FORMAT";

exports.CONFIG_SAVING_FAILED = "CONFIG_SAVING_FAILED";

exports.CONFIG_DIR_DOESNT_EXIST = "CONFIG_DIR_DOESNT_EXIST";

exports.RECORDER_TARGET_FIXTURE_DOESNT_EXISTS = "RECORDER_TARGET_FIXTURE_DOESNT_EXISTS";

exports.RECORDER_RECORDING_DOESNT_EXISTS = "RECORDER_RECORDING_DOESNT_EXISTS";

exports.RECORDER_RECORDING_IS_EMPTY = "RECORDER_RECORDING_IS_EMPTY";

exports.RECORDER_TEST_NAME_IS_EMPTY = "RECORDER_TEST_NAME_IS_EMPTY";

exports.SUITE_CFG_FILE_PARSING_FAILED = "SUITE_CFG_FILE_PARSING_FAILED";

exports.SUITE_CFG_INVALID_PROPERTY_FORMAT = "SUITE_CFG_INVALID_PROPERTY_FORMAT";

exports.SUITE_CFG_MODULE_IS_NOT_ARRAY = "SUITE_CFG_MODULE_IS_NOT_ARRAY";

exports.SUITE_CFG_PROPERTY_IS_NOT_STRING = "SUITE_CFG_PROPERTY_IS_NOT_STRING";

exports.SUITE_CFG_UNSUPPORTED_PROPERTY = "SUITE_CFG_UNSUPPORTED_PROPERTY";

exports.SUITE_GET_FILE_STATS_FAILED = "SUITE_GET_FILE_STATS_FAILED";

exports.SUITE_READ_CFG_FILE_FAILED = "SUITE_READ_CFG_FILE_FAILED";

exports.SUITE_READ_DIR_FAILED = "SUITE_READ_DIR_FAILED";

exports.TEST_RUNNER_MAX_TEST_RUN_RESTARTS = "TEST_RUNNER_MAX_TEST_RUN_RESTARTS";

exports.TEST_RUNNER_RESOURCE_ERROR_HTTP_STATUS_CODE = "TEST_RUNNER_RESOURCE_ERROR_HTTP_STATUS_CODE";

exports.VIRTUAL_FILE_SYSTEM_BASE_PATH_DOESNT_EXISTS = "VIRTUAL_FILE_SYSTEM_BASE_PATH_DOESNT_EXISTS";

exports.VIRTUAL_FILE_SYSTEM_BASE_PATH_IS_NOT_DIR = "VIRTUAL_FILE_SYSTEM_BASE_PATH_IS_NOT_DIR";

exports.VIRTUAL_FILE_SYSTEM_CREATE_DIR_FAILED = "VIRTUAL_FILE_SYSTEM_CREATE_DIR_FAILED";

exports.VIRTUAL_FILE_SYSTEM_RENAME_DIR_FAILED = "VIRTUAL_FILE_SYSTEM_RENAME_DIR_FAILED";

exports.VIRTUAL_FILE_SYSTEM_DELETE_DIR_FAILED = "VIRTUAL_FILE_SYSTEM_DELETE_DIR_FAILED";

exports.VIRTUAL_FILE_SYSTEM_INVALID_DIR_PATH = "VIRTUAL_FILE_SYSTEM_INVALID_DIR_PATH";

exports.VIRTUAL_FILE_SYSTEM_DIR_ALREADY_EXISTS = "VIRTUAL_FILE_SYSTEM_DIR_ALREADY_EXISTS";

exports.VIRTUAL_FILE_SYSTEM_DIR_DOESNT_EXIST = "VIRTUAL_FILE_SYSTEM_ERR_RENAME_DIR_PATH_NOT_EXISTS";

exports.VIRTUAL_FILE_SYSTEM_INVALID_FIXTURE_PATH = "VIRTUAL_FILE_SYSTEM_INVALID_FIXTURE_PATH";

exports.VIRTUAL_FILE_SYSTEM_FIXTURE_FILE_ALREADY_EXISTS = "VIRTUAL_FILE_SYSTEM_FIXTURE_FILE_ALREADY_EXISTS";

exports.VIRTUAL_FILE_SYSTEM_FIXTURE_FILE_DOESNT_EXIST = "VIRTUAL_FILE_SYSTEM_FIXTURE_FILE_DOESNT_EXIST";

exports.VIRTUAL_FILE_SYSTEM_RENAME_FIXTURE_FILE_FAILED = "VIRTUAL_FILE_SYSTEM_RENAME_FIXTURE_FILE_FAILED";

exports.VIRTUAL_FILE_SYSTEM_DELETE_FIXTURE_FILE_FAILED = "VIRTUAL_FILE_SYSTEM_DELETE_FIXTURE_FILE_FAILED";

exports.VIRTUAL_FILE_SYSTEM_DIR_DOESNT_READ = "VIRTUAL_FILE_SYSTEM_DIR_DOESNT_READ";

exports.VIRTUAL_FILE_SYSTEM_NODEJS_CANT_GET_DRIVE_LIST = "VIRTUAL_FILE_SYSTEM_NODEJS_CANT_GET_DRIVE_LIST";

exports.VIRTUAL_FILE_SYSTEM_READ_FIXTURE_FAILED = "VIRTUAL_FILE_SYSTEM_READ_FIXTURE_FAILED";

exports.VIRTUAL_FILE_SYSTEM_WRITE_FIXTURE_FAILED = "VIRTUAL_FILE_SYSTEM_WRITE_FIXTURE_FAILED";

exports.WORKER_POOL_FAILED_TO_START_BROWSER = "WORKER_POOL_FAILED_TO_START_BROWSER";

exports.WORKER_POOL_FAILED_TO_CREATE_WORKER_IN_BROWSER_WINDOW = "WORKER_POOL_FAILED_TO_CREATE_WORKER_IN_BROWSER_WINDOW";

exports.WORKER_POOL_WORKER_DISCONNECTED = "WORKER_POOL_WORKER_DISCONNECTED";

exports.WORKER_POOL_WORKER_NAME_IS_ALREADY_USED = "WORKER_POOL_WORKER_NAME_IS_ALREADY_USED";

exports.WORKER_POOL_WORKER_NAME_IS_EMPTY = "WORKER_POOL_WORKER_NAME_IS_EMPTY";

exports.CLI_TIMEOUT_EXCEED = "CLI_TIMEOUT_EXCEED";

exports.CLI_UNKNOWN_OPTION = "CLI_UNKNOWN_OPTION";

exports.CLI_SETUP_EXEC_NOT_EXISTS = "CLI_SETUP_EXEC_NOT_EXISTS";

exports.CLI_SETUP_ENV_PATH_NOT_DEFINED = "CLI_SETUP_ENV_PATH_NOT_DEFINED";

exports.CLI_TIMEOUT_VALUE_NOT_VALID = "CLI_TIMEOUT_VALUE_NOT_VALID";

exports.CLI_HELP_UNKNOWN_OPTION = "CLI_HELP_UNKNOWN_OPTION";

exports.CLI_CONFIG_FILE_NOT_EXISTS = "CLI_CONFIG_FILE_NOT_EXISTS";

exports.CLI_CONFIG_NOT_VALID = "CLI_CONFIG_NOT_VALID";

exports.CLI_BROWSER_PARAM_NOT_VALID = "CLI_BROWSER_PARAM_NOT_VALID";
