// testcafe/_compiled_/studio/js/components/toolbar/index.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits').default;

var _classCallCheck = require('babel-runtime/helpers/class-call-check').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _toolbarBtn = require('../toolbar-btn');

var _toolbarBtn2 = _interopRequireDefault(_toolbarBtn);

var _toolbarSeparator = require('../toolbar-separator');

var _toolbarSeparator2 = _interopRequireDefault(_toolbarSeparator);

var Toolbar = (function (_React$Component) {
    function Toolbar() {
        _classCallCheck(this, Toolbar);

        _React$Component.call(this);
    }

    _inherits(Toolbar, _React$Component);

    Toolbar.prototype.render = function render() {
        var props = {
            b1: { className: 'b1' },
            b2: { className: 'b2' },
            b3: { className: 'b3' },
            b4: { className: 'b4' },
            b5: { className: 'b5' },
            b6: { className: 'b6' },
            b7: { className: 'b7' },
            b8: { className: 'b8' }
        };

        return React.createElement(
            'ul',
            { className: 'tc-toolbar' },
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'new-project' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'open-project' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'save' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarSeparator2.default, null)
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'copy' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'paste' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarSeparator2.default, null)
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'create-fixture' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'edit-fixture-code' })
            ),
            React.createElement(
                'li',
                null,
                React.createElement(_toolbarBtn2.default, { name: 'edit-fixture' })
            )
        );
    };

    return Toolbar;
})(React.Component);

exports.default = Toolbar;
;
module.exports = exports.default;
