// testcafe/_compiled_/studio/js/components/context-menu-item/index.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits').default;

var _classCallCheck = require('babel-runtime/helpers/class-call-check').default;

exports.__esModule = true;

var MenuItem = (function (_React$Component) {
    function MenuItem() {
        _classCallCheck(this, MenuItem);

        _React$Component.call(this);

        this.state = {
            isSelected: false
        };
    }

    _inherits(MenuItem, _React$Component);

    MenuItem.prototype.render = function render() {
        var _this = this;

        var itemClass = 'tc-menu-item';

        if (this.state.isSelected) itemClass += ' selected';

        return React.createElement(
            'div',
            { className: itemClass,
                onMouseOver: function () {
                    return _this.setState({ isSelected: true });
                },
                onMouseOut: function () {
                    return _this.setState({ isSelected: false });
                },
                onClick: this.props.onClick },
            this.props.name
        );
    };

    return MenuItem;
})(React.Component);

exports.default = MenuItem;
module.exports = exports.default;
