// testcafe/_compiled_/studio/js/actions/index.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;
exports.toggleTreeItemExpand = toggleTreeItemExpand;
exports.selectTreeItem = selectTreeItem;
exports.showItemView = showItemView;
exports.closeTab = closeTab;
exports.showContextMenu = showContextMenu;

var _action = require('./action');

var _action2 = _interopRequireDefault(_action);

var _ipc = require('ipc');

var _ipc2 = _interopRequireDefault(_ipc);

function dispatch(data) {
    _ipc2.default.send('dispatch', data);
}

function toggleTreeItemExpand(node) {
    dispatch({
        actionType: _action2.default.toggleTreeItemExpand,
        node: node
    });
}

function selectTreeItem(node) {
    dispatch({
        actionType: _action2.default.selectTreeItem,
        node: node
    });
}

function showItemView(node) {
    dispatch({
        actionType: _action2.default.showItemView,
        node: node
    });
}

function closeTab(node) {
    dispatch({
        actionType: _action2.default.closeTab,
        node: node
    });
}

function showContextMenu(nodeId, mousePosition) {
    dispatch({
        actionType: _action2.default.showContextMenu,
        nodeId: nodeId,
        mousePosition: mousePosition
    });
}
