// testcafe-browser-natives/lib/index.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _apiGetInstallations = require('./api/get-installations');

var _apiGetInstallations2 = _interopRequireDefault(_apiGetInstallations);

var _apiOpen = require('./api/open');

var _apiOpen2 = _interopRequireDefault(_apiOpen);

var _apiClose = require('./api/close');

var _apiClose2 = _interopRequireDefault(_apiClose);

var _apiResize = require('./api/resize');

var _apiResize2 = _interopRequireDefault(_apiResize);

var _apiScreenshot = require('./api/screenshot');

var _apiScreenshot2 = _interopRequireDefault(_apiScreenshot);

var _apiGetBrowserInfo = require('./api/get-browser-info');

var _apiGetBrowserInfo2 = _interopRequireDefault(_apiGetBrowserInfo);

exports.default = {
    getInstallations: _apiGetInstallations2.default,
    getBrowserInfo: _apiGetBrowserInfo2.default,
    open: _apiOpen2.default,
    close: _apiClose2.default,
    resize: _apiResize2.default,
    screenshot: _apiScreenshot2.default
};
module.exports = exports.default;
