// testcafe/_compiled_/studio/js/components/tabs-view/index.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits').default;

var _classCallCheck = require('babel-runtime/helpers/class-call-check').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _tabs = require('../tabs');

var _tabs2 = _interopRequireDefault(_tabs);

var _itemView = require('../item-view');

var _itemView2 = _interopRequireDefault(_itemView);

var _emptyItemView = require('../empty-item-view');

var _emptyItemView2 = _interopRequireDefault(_emptyItemView);

var _ipc = require('ipc');

var _ipc2 = _interopRequireDefault(_ipc);

var TabsView = (function (_React$Component) {
    function TabsView() {
        _classCallCheck(this, TabsView);

        _React$Component.call(this);

        this.state = {
            tabs: [],
            activeId: ''
        };
    }

    _inherits(TabsView, _React$Component);

    TabsView.prototype.componentDidMount = function componentDidMount() {
        _ipc2.default.on('tabsViewChange', this._onChange.bind(this));
    };

    TabsView.prototype.componentWillUnmount = function componentWillUnmount() {
        _ipc2.default.removeListener('tabsViewChange', this._onChange.bind(this));
    };

    TabsView.prototype.render = function render() {
        var _this = this;

        var activeView = this.state.tabs.filter(function (tab) {
            return tab.id === _this.state.activeId;
        })[0];

        return activeView ? React.createElement(
            'div',
            { className: 'tc-tabs-view' },
            React.createElement(_tabs2.default, { tabs: this.state.tabs, activeId: this.state.activeId }),
            React.createElement(_itemView2.default, { key: this.state.activeId, node: activeView })
        ) : React.createElement(_emptyItemView2.default, null);
    };

    TabsView.prototype._onChange = function _onChange(state) {
        this.setState(state);
    };

    return TabsView;
})(React.Component);

exports.default = TabsView;
module.exports = exports.default;
